wireapp/wire-android
# chore: replace current navigation with Voyager [AR-2012]

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/AR-2012" title="AR-2012" target="_blank"><img alt="Task" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10818?size=medium" />AR-2012</a>  [Spike] Navigation issues -> Fix
  </td></table>
  <br />
 






----
#### PR Submission Checklist for internal contributors

- The **PR Title**

- The **PR Description**
----

# What's new in this PR?

### Issues

Currently, we have some issues related to the navigation, like: typesafety, passing arguments forth and back, deeplink handling, transitions, overcomplicated code.

### Solutions

This is just a proof of concept so there still may be things that don’t work properly, also tests doesn’t fully pass, because I focused on checking how it would work in our app, not implementing the whole thing as it should be.

Hilt with ViewModel doesn’t support AssistedInject, and that’s one of the main points of using this library (typesafety), on the other hand library offers a substitute in a form of ScreenModel, but unfortunately it bypasses the whole ViewModelComponent and thus we couldn’t keep using ViewModel scope, and it’s used a lot in our app. The chosen solution works with the use of SavedStateHandle, which is the only parameter allowed by Hilt to be passed by AssistedInject-ish way, so there is a new method `getAssistedViewModel` (instead of standard `hiltViewModel` Voyager lib has its own `getViewModel` which works with their `Screen` approach), this new method takes `param` which can be of `Any` type and it’s injected into ViewModel using SavedStateHandle. Passing arguments back is easy - using simply lambdas.
All NavHosts and standard NavGraphControllers are replaced with Navigators, there was no way of combining both in the same app (too many crashes and issues, it was easier to just replace all). Some NavHosts, mainly the ones that shouldn’t be created as a navigation elements in the first place, are replaced by way simpler solutions. For instance, when we have just some composables that should be replaced depending on the state, it’s exactly how Crossfade works, so there is no need to introduce another navigation level.
Handling deep links also is simplified, instead of complex string navigation attributes, there is just a sealed class with current deep link state and everything is handled in the one place.

### Testing

#### Test Coverage (Optional)


Automated tests are not yet adjusted at this point, there is no sense in spending time before deciding whether we want such a solution at all.

#### How to Test

Just build and open the app and start navigating to different screens to see if everything is correct.

----
#### PR Post Submission Checklist for internal contributors (Optional)

----
#### PR Post Merge Checklist for internal contributors

----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.
